rasterio/_base.pyx

```diff
@@ -202,7 +202,7 @@ cdef class DatasetBase(object):
         self._hds = NULL
 
         if path is not None:
-            filename = vsi_path(path)
+            filename = parse_path(path).as_vsi()
```

Comment:
The new as_vsi method does the same thing as before and now we also fail right away if path is the wrong type instead of maybe succeeding in some cases.